pkg/kubelet/cm/deviceplugin/manager.go

```diff
@@ -283,46 +251,54 @@ func (m *ManagerImpl) Allocate(node *schedulercache.NodeInfo, attrs *lifecycle.P
 }
 
 // Register registers a device plugin.
-func (m *ManagerImpl) Register(ctx context.Context, r *pluginapi.RegisterRequest) (*pluginapi.Empty, error) {
-	glog.Infof("Got registration request from device plugin with resource name %q", r.ResourceName)
-	metrics.DevicePluginRegistrationCount.WithLabelValues(r.ResourceName).Inc()
-	if r.Version != pluginapi.Version {
-		errorString := fmt.Sprintf(errUnsupportedVersion, r.Version, pluginapi.Version)
-		glog.Infof("Bad registration request from device plugin with resource name %q: %v", r.ResourceName, errorString)
-		return &pluginapi.Empty{}, fmt.Errorf(errorString)
+func (m *ManagerImpl) register(r *pluginwatcher.PluginInfo) {
+	name, err := pluginwatcher.GetResourceNameFrom(r.SockPath)
+	if err != nil {
+		glog.Infof("Invalid name of device plugin socket: %v", err)
+		return
 	}
+	glog.Infof("Calling registration request to device plugin with resource name %q", name)
+	metrics.DevicePluginRegistrationCount.WithLabelValues(name).Inc()
 
-	if !v1helper.IsExtendedResourceName(v1.ResourceName(r.ResourceName)) {
-		errorString := fmt.Sprintf(errInvalidResourceName, r.ResourceName)
-		glog.Infof("Bad registration request from device plugin: %v", errorString)
-		return &pluginapi.Empty{}, fmt.Errorf(errorString)
+	if !v1helper.IsExtendedResourceName(v1.ResourceName(name)) {
+		errorString := fmt.Sprintf(errInvalidResourceName, name)
+		glog.Infof("Invalid name of device plugin socket: %v", errorString)
+		return
+	}
+	conn, err := pluginwatcher.Dial(r.SockPath)
+	if err != nil {
+		glog.Errorf("Failed to register device plugin with request %v: %v", r, err)
+		// TODO: cleanup. os.RemoveAll(r.SockPath)
+		return
 	}
 
-	// TODO: for now, always accepts newest device plugin. Later may consider to
-	// add some policies here, e.g., verify whether an old device plugin with the
-	// same resource name is still alive to determine whether we want to accept
-	// the new registration.
-	go m.addEndpoint(r)
-
-	return &pluginapi.Empty{}, nil
+	existingDevs := make(map[string]pluginapi.Device)
+	e := newEndpointImpl(r.SockPath, name, conn, existingDevs, m.callback)
+	err = e.register(name)
```

Comment:
Do we expect the plugins to support register call? CSI does not support any register call today.

Reply:
We probably should re-use the service Identity of the [CSI spec](https://github.com/container-storage-interface/spec/blob/master/spec.md)